### PR TITLE
feat: add optional line range params to read_file tool

### PR DESCRIPTION
**Pull request by wwwillchen**
## Summary
- Add `start_line_one_indexed` and `end_line_one_indexed_inclusive` optional parameters to the `read_file` agent tool, allowing the agent to read specific line ranges from files
- Update consent preview to display line range (e.g., "Read src/App.tsx (lines 10-50)") and XML builder to include `start_line`/`end_line` attributes
- Add comprehensive unit tests (36 tests) with minimal mocking, using real temp files on disk

## Test plan
- [x] All 36 new unit tests pass covering: schema validation, full file reads, start-only, end-only, both params, edge cases (out-of-bounds clamping, start > end, empty files, single-line files), consent preview, and XML builder
- [x] All 747 existing tests pass (no regressions)
- [x] Lint, formatting, and type checks pass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

<a href="https://app.devin.ai/review/dyad-sh/dyad/pull/2516" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

---
## Summary by cubic
Adds optional line range parameters to the read_file tool so the agent can read specific lines and reflect the range in consent preview and XML. Improves safety with clamping, schema validation, and clear errors, backed by comprehensive tests.

- **New Features**
  - Support start_line_one_indexed and end_line_one_indexed_inclusive to read a specific line range.
  - Consent preview shows ranges; XML builder adds start_line/end_line attributes.

- **Bug Fixes**
  - Escape start_line/end_line in XML and update parser to accept dyad-read tags with attributes.
  - Validate start <= end and fix trailing newline handling to match full-file reads.

<sup>Written for commit 4c5c74afc7d7e388dfa821f21d04991e9454cb5e. Summary will update on new commits.</sup>

---

> [!NOTE]
> **Medium Risk**
> Touches agent tool I/O schema and file-reading behavior, which could subtly change what content is exposed/returned (especially newline and range edge cases), but changes are localized and heavily tested.
> 
> **Overview**
> Adds optional `start_line_one_indexed` / `end_line_one_indexed_inclusive` parameters to the local agent `read_file` tool, including validation (`start <= end`), consent preview messaging, and `<dyad-read>` XML output with `start_line`/`end_line` attributes.
> 
> Updates `read_file` execution to return only the requested line range (with clamping and trailing-newline preservation for full-file-equivalent ranges), and loosens the message parser regex to recognize `<dyad-read>` tags that include additional attributes.
> 
> Adds a comprehensive `read_file` unit test suite and updates local-agent E2E request snapshots to reflect the expanded tool schema (and an additional `store: false` request field).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 4c5c74afc7d7e388dfa821f21d04991e9454cb5e. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by wwwillchen** [comment]
@BugBot run

File: src/pro/main/ipc/handlers/local_agent/tools/read_file.ts [hunk 1]
```diff
@@ -32,6 +62,18 @@ export const readFileTool: ToolDefinition<z.infer<typeof readFileSchema>> = {
...
+    const end = args.end_line_one_indexed_inclusive;
+
+    if (start == null && end == null) {
+      return content;
+    }
+
+    const lines = content.split("\n");
+    const startIdx = Math.max(0, (start ?? 1) - 1);
+    const endIdx = Math.min(lines.length, end ?? lines.length);
+    return lines.slice(startIdx, endIdx).join("\n");
```
[the remark this answers is not in the document]
Comment (wwwillchen) [reply src/pro/main/ipc/handlers/local_agent/tools/read_file.ts]: don't boether!

**Comment by wwwillchen** [comment]
@BugBot run

**Comment by wwwillchen** [comment]
@BugBot run
